### DISCUSSION TITLE
Running npm as a Windows service

### DISCUSSION THREAD

**Discussion by ttblum** [community/community#137201] [npm]
Hello,

I have an npm app running successfully on my Windows 2022 server, launching it with command:

`npm run start:production`

I configured it to run as a Windows service using NSSM ( https://github.com/treeswift/nssm ) , NSSM is set up to launch:

`"C:\Program Files\nodejs\npm.cmd" run start:production`

The app launches OK this way, but users complain that they aren't able to authenticate to external website APIs, as if it wasn't reading the API keys correctly that are in the .env file.

Is it possible to launch npm apps as a Windows service?

If so, what is the best way to run npm apps as a Windows service, using NSSM or some other method?  Should I be launching a .bat file, or running npm with arguments, running as LOCALSYSTEM or as a local user, etc.?

**Comment by Gogo200202** [comment]
Maybe you don’t have permission to open this file

**Comment by tshandler** [comment]
This will be helpful of your service.

1. Ensure the .env file is accessible and readable by the service.
2. Pass environment variables manually via NSSM or ensure they are get globally.
3. Try running the service as a local user instead of LOCALSYSTEM.
4. Launch the app via a .bat file to ensure proper loading.
5. Consider switching to PM2 for better process management.

These steps should help ensure your app can access the .env file and function properly when running as a Windows service. 
Let me know if you need further assistance!

**Comment by ttblum** [comment]
Thanks.

I have tried setting NSSM to launch a .bat file and also tried launching with:

`"C:\Program Files\nodejs\npm.cmd" run start:production`

I have tried setting the service to run as LOCALSYSTEM, and have tried running it as a domain user. I have tried pasting the contents of the .env file into NSSM as environment variables.

I will try setting the service to run as a local user.

How would I launch it with PM2? Would it be:

`pm2 start 'npm run start:production'`

?

**Reply by ttblum** [reply]
Running the service as LOCALSYSTEM did not work, running the service as a domain user did not work. The app launches but does not read the .env file correctly.

Running it as a .bat file as a local Windows account works. The app runs successfully this way.
